File: Hash.cpp
#include <iostream>
#include <string>

using namespace std;

// Funcion para hacer la tabla HASH

int clave(string cadena)
{
    int maximo = cadena.length();
    int clave_dni = 0;
    for (int i = 0; i < maximo; i++)
    {
        clave_dni = clave_dni + int(cadena[i]) - int('0');
    }
    return clave_dni;
}

void Hash(string L[][2], int cant)
{
    string Copia[cant*2];
    int cont=0;
    for(int i=0;i<cant;i++){
        for(int j=0;j<2;j++){
            Copia[cont]=L[i][j];
            cont+=1;
        }
    }
    for (int i = 0; i < cant; i++)
    {
        L[i][0]="";
        L[i][1]="";
    }

    for(int i=0;i<cant*2;i+=2){
        int codigo = clave(Copia[i]);
        int indice = codigo % cant;
        while(true){
            indice = indice % cant;
            if(L[indice][0]==""){
                L[indice][0]=Copia[i];
                L[indice][1]=Copia[i+1];
                break;
            }
            else{
                indice+=1;
            }
        }

    }


    
}

void menu(string L[][2],int cant){
    cout<<"\nMenu\n";
    cout<<"1. Consultar DNI\n2. Eliminar DNI\n3. Modificar DNI\n4. Salir\n";
    int valor;
    cin>>valor;
    string dni;
    int clave_dni;
    int max;
    int opc=1;
    
    if(valor == 4){
        cout<<"Adios\n";
    }

    else{
        cout<<"Ingrese su DNI: ";
        switch (valor)
        {
        case 1:
            cin>>ws;
            getline(cin,dni);
            cout<<dni<<endl;
            clave_dni = clave(dni);
            max = clave_dni;
            clave_dni%=cant;
            while(dni != L[clave_dni][0]){
                clave_dni++;
                clave_dni%=cant;
                if(clave_dni==max)
                {
                    break;
                    opc = -1;
                }
            }
            if(opc == 1){
                cout<<"La persona que tiene el dni "<<dni<<" es: "<<L[clave_dni][1]<<endl;
            }
            else{
                cout<<"No se encontro el dni\n";
            }
            break;
        
    

            default:
                cout<<"Ingrese otro valor";
                break;
        }
    }
}

void mostrar(string L[][2], int cant)
{
    cout << "Indice\tDNI\t\tNombre\n";
    for (int i = 0; i < cant; i++)
    {
        cout << i << "\t" << L[i][0] << "\t" << L[i][1] << endl;
    }
}

int main()
{
    string L[][2] = {{"12345678", "Juan"},
                     {"96854326", "María"},
                     {"85215647", "Rosa"},
                     {"50247895", "Pablo"},
                     {"23659874", "Federico"},
                     {"86597545", "Guadalupe"},
                     {"23146385", "Benito"},
                     {"89756413", "Carolina"},
                     {"96445646", "Charlie"},
                     {"74520367", "Donald"}};
    int cant = sizeof(L) / sizeof(L[0]);
    mostrar(L, cant);
    Hash(L,cant);
    cout<<"\n\n\n";
    mostrar(L,cant);
    menu(L,cant);
}

File: Hash.py

def Orden(L):
    # Pasamos los dni a una tupla
    tupla = []
    for i in range(len(L)):
        tupla += [L[i][0], L[i][1]]
    tupla = tuple(tupla)
    for i in range(len(L)):
        L[i][0] = ""
        L[i][1] = ""

    # tupla = ('1213141','asf','321321312','fddsf',...)

    lista = []
    for i in range(0, len(tupla), 2):
        dni = list(tupla[i])
        clave = 0
        for j in dni:
            clave += int(j)
        indice = clave % len(L)
        # Se empieza con Juan entonces el indice sera = 6
        lista += [tupla[i]]
        while True:
            indice = indice % len(L)
            if L[indice][0] == "":
                L[indice][0] = tupla[i]
                L[indice][1] = tupla[i+1]
                break
            else:
                indice += 1


def table(L):
    print("ÍNDICE\t  DNI\t\tNOMBRE")
    for i in range(len(L)):
        print(i, "\t", L[i][0], "\t", L[i][1])


def Menu(L):
    valor = 0
    while(valor != 4):
        print("\nMenú\n")
        print("1. Consultar por el DNI")
        print("2. Cambiar registro con DNI")
        print("3. Eliminar DNI")
        print("4. Salir")
        valor = int(input("Escriba la opcion: "))
        if valor == 4:
            break
        consulta = input("Ingrese el DNI: ")
        clave = 0
        for i in list(consulta):
            clave += int(i)
        clave = clave % len(L)
        maximo = clave
        if valor == 1:
            if consulta == L[clave][0]:
                indice = clave
            else:
                while consulta != L[clave % len(L)][0]:
                    clave += 1
                    if(clave == maximo + len(L)):
                        clave = -1
                        break
                indice = clave
            if indice == -1:
                print("No se encontro ese DNI, lo sentimos")
            else:
                print("La persona con ese DNI es: ", L[indice][1])
        elif valor == 2:
            if consulta == L[clave][0]:
                indice = clave
            else:
                while consulta != L[clave % len(L)][0]:
                    clave += 1
                    if(clave == maximo + len(L)):
                        clave = -1
                        break
                indice = clave
            if indice == -1:
                print("No se encontro ese DNI, lo sentimos")
            else:
                Nuevo_DNI = input("Escriba su nuevo DNI: ")
                Nuevo_Nombre = input("Escriba su nuevo nombre: ")
                L[indice][0], L[indice][1] = Nuevo_DNI, Nuevo_Nombre
                Orden(L)
                table(L)
        elif valor == 3:
            if consulta == L[clave][0]:
                indice = clave
            else:
                while consulta != L[clave % len(L)][0]:
                    clave += 1
                    if(clave == maximo + len(L)):
                        clave = -1
                        break
                indice = clave
            if indice == -1:
                print("No se encontro ese DNI, lo sentimos")
            else:
                L.remove(L[indice])
                Orden(L)
                table(L)
        print("\n")
        # Si se trabajara con n elementos


def Lista_DefinidaporUser():
    lista = []
    cant = int(input("Cantidad de personas: "))
    for i in range(cant):
        DNI = input("Ingrese su numero de DNI: ")
        Nombre = input("Ingrese su nombre: ")
        lista.append([DNI, Nombre])
    return lista


L = [['12345678', 'Juan'], ['96854326', 'María'],
     ['85215647', 'Rosa'], ['50247895', 'Pablo'], ['23659874', 'Federico'], ['86597545', 'Guadalupe'], ['23146385', 'Benito'], ['89756413', 'Carolina'], ['96445646', 'Charlie'], ['74520367', 'Donald']]

# Ordenarlo por HASH
Orden(L)
# Mostrarlo
table(L)
# MENU
print("\n")
Menu(L)

# Definidas por el usuario
print("\n\n Lista Definida por el Usuario\n")
lista_def = Lista_DefinidaporUser()
Orden(lista_def)
table(lista_def)
Menu(lista_def)
